File: app/src/main/java/com/myjb/dev/network/PriceInquiry.java
package com.myjb.dev.network;

import android.os.AsyncTask;
import android.support.annotation.NonNull;
import android.util.Log;

import com.myjb.dev.model.ServiceModel;
import com.myjb.dev.mygaragesale.BuildConfig;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

import java.io.IOException;
import java.io.UnsupportedEncodingException;
import java.net.MalformedURLException;
import java.net.URLEncoder;
import java.net.UnknownHostException;
import java.util.ArrayList;
import java.util.List;

public class PriceInquiry extends AsyncTask<Void, Void, List<BookInfoItem>> {

    public interface OnPriceListener {
        void onPriceResult(List<BookInfoItem> priceList);
    }

    protected final static String TAG = "PriceInquiry";

    protected String baseUrl;
    private String query;
    private OnPriceListener listener;

    public PriceInquiry(@NonNull String query, @NonNull OnPriceListener listener) {
        this.query = query;
        this.listener = listener;
    }

    @Override
    protected List<BookInfoItem> doInBackground(Void... params) {
        String url = null;
        try {
            url = baseUrl + URLEncoder.encode(query, "EUC-KR");
        } catch (UnsupportedEncodingException e) {
            e.printStackTrace();
        }

        if (BuildConfig.DEBUG)
            Log.e(TAG, "[basicVersion] url : " + url);

        return getPriceInfo(url);
    }

    @Override
    protected void onPostExecute(List<BookInfoItem> imageUrls) {
        if (listener != null)
            listener.onPriceResult(imageUrls);
    }

    protected List<BookInfoItem> getPriceInfo(String url) {
        Document doc = null;
        Elements elements = new Elements();

        try {
            long init = System.currentTimeMillis();

            doc = getDocument(url);

            long connect = System.currentTimeMillis();

            elements = getElements(doc);

            long select = System.currentTimeMillis();

            List<BookInfoItem> bookInfoList = getItems(elements);

            if (BuildConfig.DEBUG)
                Log.e(TAG, "[basicVersion] connect : " + (connect - init) + ", select : " + (select - connect) + ", add : " + (System.currentTimeMillis() - select));

            return bookInfoList;
        } catch (MalformedURLException e) {
            e.printStackTrace();
        } catch (UnknownHostException e) {
            e.printStackTrace();
        } catch (IOException e) {
            e.printStackTrace();
        } finally {
            if (elements != null)
                elements = null;
            if (doc != null)
                doc = null;
        }

        return null;
    }

    private Document getDocument(String url) throws IOException {
        return Jsoup.connect(url).get();
    }

    @NonNull
    protected Elements getElements(Document doc) {
        String filter = getBasicFilter();

        if (BuildConfig.DEBUG)
            Log.e(TAG, "[basicVersion] filter : " + filter);

        return doc.select(filter);
    }

    @NonNull
    private List<BookInfoItem> getItems(Elements elements) {
        List<BookInfoItem> bookInfoList = new ArrayList<>();
        for (Element element : elements) {
            String isbn = getISBN(element);

            if (BuildConfig.DEBUG)
                Log.e(TAG, "[basicVersion] isbn : " + isbn);

            String image = element.select(getImageFilter()).attr("src");

            if (BuildConfig.DEBUG)
                Log.e(TAG, "[basicVersion] getImageFilter() : " + getImageFilter() + ", image : " + image);

            String name = element.select(getNameFilter()).text();

            if (BuildConfig.DEBUG)
                Log.e(TAG, "[basicVersion] getNameFilter() : " + getNameFilter() + ", name : " + name);

            String[] price = element.select(getPriceFilter()).text().split("원");
            if (price.length > 1)
                bookInfoList.add(new BookInfoItem(isbn, getCompany(), image, name, new PriceItem(price)));
        }

        if (BuildConfig.DEBUG)
            Log.e(TAG, "[basicVersion] list.size : " + bookInfoList.size());

        return bookInfoList;
    }

    protected int getCompany() {
        return ServiceModel.Company.NONE;
    }

    @NonNull
    protected String getBasicFilter() {
        return "";
    }

    @NonNull
    protected String getISBN(Element element) {
        return "";
    }

    @NonNull
    protected String getImageFilter() {
        return "img[abs:src]";
    }

    @NonNull
    protected String getNameFilter() {
        return "";
    }

    @NonNull
    protected String getPriceFilter() {
        return "";
    }
}

File: app/src/main/java/com/myjb/dev/network/PriceInquiry.kt
package com.myjb.dev.network

import android.os.AsyncTask
import android.util.Log
import com.myjb.dev.model.ServiceModel
import com.myjb.dev.mygaragesale.BuildConfig
import org.jsoup.Jsoup
import org.jsoup.nodes.Document
import org.jsoup.nodes.Element
import org.jsoup.select.Elements
import java.io.IOException
import java.io.UnsupportedEncodingException
import java.net.MalformedURLException
import java.net.URLEncoder
import java.net.UnknownHostException

open class PriceInquiry(private val query: String, listener: OnPriceListener) :
    AsyncTask<Void?, Void?, List<BookInfoItem>?>() {
    interface OnPriceListener {
        fun onPriceResult(priceList: List<BookInfoItem>?)
    }

    protected var baseUrl: String? = null
    private val listener: OnPriceListener? = listener

    protected override fun doInBackground(vararg params: Void?): List<BookInfoItem>? {
        var url: String? = null
        try {
            url = baseUrl + URLEncoder.encode(query, "UTF-8")
        } catch (e: UnsupportedEncodingException) {
            e.printStackTrace()
        }

        if (BuildConfig.DEBUG) Log.e(TAG, "[basicVersion] url : $url")

        return getPriceInfo(url)
    }

    override fun onPostExecute(imageUrls: List<BookInfoItem>?) {
        listener?.onPriceResult(imageUrls)
    }

    protected fun getPriceInfo(url: String?): List<BookInfoItem>? {
        var doc: Document? = null
        var elements: Elements? = Elements()

        try {
            val init = System.currentTimeMillis()

            doc = getDocument(url)

            val connect = System.currentTimeMillis()

            elements = getElements(doc)

            val select = System.currentTimeMillis()

            val bookInfoList = getItems(elements)

            if (BuildConfig.DEBUG) Log.e(
                TAG,
                "[basicVersion] connect : " + (connect - init) + ", select : " + (select - connect) + ", add : " + (System.currentTimeMillis() - select)
            )

            return bookInfoList
        } catch (e: MalformedURLException) {
            e.printStackTrace()
        } catch (e: UnknownHostException) {
            e.printStackTrace()
        } catch (e: IOException) {
            e.printStackTrace()
        } finally {
            if (elements != null) elements = null
            if (doc != null) doc = null
        }

        return null
    }

    @Throws(IOException::class)
    private fun getDocument(url: String?): Document {
        return Jsoup.connect(url).get()
    }

    protected open fun getElements(doc: Document?): Elements {
        val filter = basicFilter

        if (BuildConfig.DEBUG) Log.e(TAG, "[basicVersion] filter : $filter")

        return doc!!.select(filter)
    }

    private fun getItems(elements: Elements): List<BookInfoItem> {
        val bookInfoList: MutableList<BookInfoItem> = ArrayList()
        for (element in elements) {
            val isbn = getISBN(element)

            if (BuildConfig.DEBUG) Log.e(TAG, "[basicVersion] isbn : $isbn")

            val image = element.select(imageFilter).attr("src")

            if (BuildConfig.DEBUG) Log.e(
                TAG,
                "[basicVersion] getImageFilter() : " + imageFilter + ", image : " + image
            )

            val name = element.select(nameFilter).text()

            if (BuildConfig.DEBUG) Log.e(
                TAG,
                "[basicVersion] getNameFilter() : " + nameFilter + ", name : " + name
            )

            val price = element.select(priceFilter).text().split("원".toRegex())
                .dropLastWhile { it.isEmpty() }
                .toTypedArray()
            if (price.size > 1) bookInfoList.add(
                BookInfoItem(
                    isbn,
                    company,
                    image,
                    name,
                    PriceItem(price)
                )
            )
        }

        if (BuildConfig.DEBUG) Log.e(TAG, "[basicVersion] list.size : " + bookInfoList.size)

        return bookInfoList
    }

    protected open val company: Int
        get() = ServiceModel.Company.NONE

    protected open val basicFilter: String
        get() = ""

    protected open fun getISBN(element: Element): String {
        return ""
    }

    protected val imageFilter: String
        get() = "img[abs:src]"

    protected open val nameFilter: String
        get() = ""

    protected open val priceFilter: String
        get() = ""

    companion object {
        protected const val TAG: String = "PriceInquiry"
    }
}
